asax/calculator.py

```diff
@@ -29,46 +29,34 @@ def update(self, atoms: Atoms):
             raise RuntimeError("Need an Atoms object to do anything!")
 
         if self.atoms is None:
-            self.atoms = atoms.copy()
+            # new or reset calculator: redo everything
+            self.on_atoms_changed(atoms)
+            self.atoms = atoms.copy()  # avoid missing changes
             self.results = {}
-            self.on_atoms_changed()
             self.setup()
             return
 
         changes = compare_atoms(self.atoms, atoms)
-        if not changes:
-            return
-
-        # cache not empty and we got a new atom that has changes
-        # => clear results, clear cache, re-run update function to write new atom to cache
-        self.results = {}
-        if "cell" in changes:
-            # TODO: Does this include switches from bulk to molecules?
-            # => displacement only requires re-initialization if this is the case
-            self.atoms = None
-            self.update(atoms)
-            return
-
-        if "positions" in changes:
-            self.results = self.compute_properties()
-            return
+        if changes:
+            # any change invalidates results
+            self.results = {}
 
-        # TODO: Detect changes in atom count/shape
-        # => potential only requires re-initialization if this is the case
+            if "cell" in changes or "pbc" in changes:
+                # need to recreate displacement; might as well start over
+                self.atoms = None
+                self.update(atoms)
```

Comment:
Isn't there a `return` missing here? The current implementation causes two calls of `self.on_atoms_changed()` for cell or PBC changes. See the following commit with a failing test case for this.

Reply:
> Isn't there a `return` missing here? The current implementation causes two calls of `self.on_atoms_changed()` for cell or PBC changes. See the following commit with a failing test case for this.

Hm, I don't understand what's going on there. I thought the call to `self.update()` would stop execution here and start from the top...? Strange.


> (Running the test cases individually doesn't expose this error, which is usually a hint that the unit tests are not atomic. I'd move the calculator initializations to `setUp()` to prevent stateful behavior here).

Why are there two tests for LJ anyway? It makes more sense in the `setUp()` for sure, that's what it's for... did I originally write this? If yes, shame on me.